File: backend/internal/models/campaign.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ── Enum types ──────────────────────────────────────────────────────────────

type CampaignStatus string

const (
	CampaignStatusDraft     CampaignStatus = "draft"
	CampaignStatusActive    CampaignStatus = "active"
	CampaignStatusPaused    CampaignStatus = "paused"
	CampaignStatusCompleted CampaignStatus = "completed"
	CampaignStatusCanceled  CampaignStatus = "canceled"
)

type CampaignLeadStatus string

const (
	CampaignLeadStatusPending    CampaignLeadStatus = "pending"
	CampaignLeadStatusInProgress CampaignLeadStatus = "in_progress"
	CampaignLeadStatusCompleted  CampaignLeadStatus = "completed"
	CampaignLeadStatusPaused     CampaignLeadStatus = "paused"
	CampaignLeadStatusFailed     CampaignLeadStatus = "failed"
	CampaignLeadStatusRemoved    CampaignLeadStatus = "removed"
	CampaignLeadStatusSkipped    CampaignLeadStatus = "skipped"
)

type SequenceStepType string

const (
	StepTypeFollow            SequenceStepType = "follow"
	StepTypeLikePost          SequenceStepType = "like_post"
	StepTypeConnectionRequest SequenceStepType = "connection_request"
	StepTypeMessage           SequenceStepType = "message"
	StepTypeEmail             SequenceStepType = "email"
	StepTypeInMail            SequenceStepType = "inmail"
	StepTypeViewProfile       SequenceStepType = "view_profile"
	StepTypeDelay             SequenceStepType = "delay"
)

type ConditionType string

const (
	ConditionAccepted    ConditionType = "accepted"
	ConditionNotAccepted ConditionType = "not_accepted"
	ConditionReplied     ConditionType = "replied"
	ConditionNotReplied  ConditionType = "not_replied"
	ConditionOpened      ConditionType = "opened"
	ConditionClicked     ConditionType = "clicked"
)

type ActivityType string

const (
	ActivityFollow            ActivityType = "follow"
	ActivityUnfollow          ActivityType = "unfollow"
	ActivityLikePost          ActivityType = "like_post"
	ActivityConnectionRequest ActivityType = "connection_request"
	ActivityMessageSent       ActivityType = "message_sent"
	ActivityReplyReceived     ActivityType = "reply_received"
	ActivityInMailSent        ActivityType = "inmail_sent"
	ActivityEmailSent         ActivityType = "email_sent"
	ActivityProfileView       ActivityType = "profile_view"
)

type ActivityStatus string

const (
	ActivityStatusSuccess ActivityStatus = "success"
	ActivityStatusFailed  ActivityStatus = "failed"
	ActivityStatusPending ActivityStatus = "pending"
	ActivityStatusSkipped ActivityStatus = "skipped"
)

// ── Campaign ────────────────────────────────────────────────────────────────

type Campaign struct {
	ID          uuid.UUID      `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	UserID      uuid.UUID      `gorm:"type:uuid;not null;index" json:"user_id"`
	Name        string         `gorm:"type:text;not null" json:"name"`
	Description *string        `gorm:"type:text" json:"description,omitempty"`
	Status      CampaignStatus `gorm:"type:text;default:'draft'" json:"status"`

	// Lead list association
	LeadListID *uuid.UUID `gorm:"type:uuid" json:"lead_list_id,omitempty"`

	// Configuration
	DailyLimit int    `gorm:"default:50" json:"daily_limit"`
	Timezone   string `gorm:"type:text;default:'UTC'" json:"timezone"`

	// Safety Controls
	WorkingHoursStart       *string     `gorm:"type:text" json:"working_hours_start,omitempty"`
	WorkingHoursEnd         *string     `gorm:"type:text" json:"working_hours_end,omitempty"`
	WorkingDays             StringArray `gorm:"type:jsonb" json:"working_days,omitempty"`
	DelayMinSeconds         *int        `gorm:"default:45" json:"delay_min_seconds,omitempty"`
	DelayMaxSeconds         *int        `gorm:"default:120" json:"delay_max_seconds,omitempty"`
	WarmUpEnabled           *bool       `gorm:"default:false" json:"warm_up_enabled,omitempty"`
	WarmUpDays              *int        `gorm:"default:14" json:"warm_up_days,omitempty"`
	WarmUpStartDate         *time.Time  `gorm:"type:timestamptz" json:"warm_up_start_date,omitempty"`
	AutoPauseBelowAccept    *float64    `gorm:"column:auto_pause_below_acceptance" json:"auto_pause_below_acceptance,omitempty"`
	SkipAlreadyContacted    *bool       `gorm:"default:true" json:"skip_already_contacted,omitempty"`
	StopOnReply             *bool       `gorm:"default:true" json:"stop_on_reply,omitempty"`

	// Performance Metrics (aggregates)
	TotalLeads     int `gorm:"default:0" json:"total_leads"`
	PendingLeads   int `gorm:"default:0" json:"pending_leads"`
	CompletedLeads int `gorm:"default:0" json:"completed_leads"`
	RepliedLeads   int `gorm:"default:0" json:"replied_leads"`

	// Stats
	ConnectionSent     int `gorm:"default:0" json:"connection_sent"`
	ConnectionAccepted int `gorm:"default:0" json:"connection_accepted"`
	MessagesSent       int `gorm:"default:0" json:"messages_sent"`
	RepliesReceived    int `gorm:"default:0" json:"replies_received"`
	FailedLeads        int `gorm:"default:0" json:"failed_leads"`

	// Timestamps
	StartedAt   *time.Time `gorm:"type:timestamptz" json:"started_at,omitempty"`
	PausedAt    *time.Time `gorm:"type:timestamptz" json:"paused_at,omitempty"`
	CompletedAt *time.Time `gorm:"type:timestamptz" json:"completed_at,omitempty"`
	CreatedAt   time.Time  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time  `gorm:"autoUpdateTime" json:"updated_at"`

	// Relations (not stored in DB — populated by service layer)
	Senders   []CampaignSender   `gorm:"-" json:"senders,omitempty"`
	Sequences []CampaignSequence `gorm:"-" json:"sequences,omitempty"`
	Leads     []CampaignLead     `gorm:"-" json:"leads,omitempty"`
}

func (Campaign) TableName() string { return "campaigns" }

func (c *Campaign) BeforeCreate(tx *gorm.DB) error {
	if c.ID == uuid.Nil {
		c.ID = uuid.New()
	}
	return nil
}

// ── CampaignSender ──────────────────────────────────────────────────────────

type CampaignSender struct {
	ID                 uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	CampaignID         uuid.UUID `gorm:"type:uuid;not null;index" json:"campaign_id"`
	LinkedInAccountID  uuid.UUID `gorm:"column:linkedin_account_id;type:uuid;not null" json:"linkedin_account_id"`
	IsActive           bool      `gorm:"default:true" json:"is_active"`
	DailyLimit         int       `gorm:"default:50" json:"daily_limit"`

	// Performance Metrics
	LeadsAssigned      int `gorm:"default:0" json:"leads_assigned"`
	ConnectionSent     int `gorm:"default:0" json:"connection_sent"`
	ConnectionAccepted int `gorm:"default:0" json:"connection_accepted"`
	MessagesSent       int `gorm:"default:0" json:"messages_sent"`
	RepliesReceived    int `gorm:"default:0" json:"replies_received"`

	// Timestamps
	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	// Relations
	LinkedInAccount *LinkedInAccount `gorm:"foreignKey:LinkedInAccountID" json:"linkedin_account,omitempty"`
}

func (CampaignSender) TableName() string { return "campaign_senders" }

func (cs *CampaignSender) BeforeCreate(tx *gorm.DB) error {
	if cs.ID == uuid.Nil {
		cs.ID = uuid.New()
	}
	return nil
}

// ── CampaignSequence ────────────────────────────────────────────────────────

type CampaignSequence struct {
	ID         uuid.UUID        `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	CampaignID uuid.UUID        `gorm:"type:uuid;not null;index" json:"campaign_id"`
	StepNumber int              `gorm:"not null" json:"step_number"`
	StepType   SequenceStepType `gorm:"type:text;not null" json:"step_type"`

	// Action Details
	MessageTemplate  *string `gorm:"type:text" json:"message_template,omitempty"`
	SubjectTemplate  *string `gorm:"type:text" json:"subject_template,omitempty"`
	PostURL          *string `gorm:"type:text" json:"post_url,omitempty"`

	// A/B Testing
	MessageTemplateB *string `gorm:"type:text" json:"message_template_b,omitempty"`
	ABTestEnabled    *bool   `gorm:"default:false" json:"ab_test_enabled,omitempty"`
	ABTestWinner     *string `gorm:"type:text" json:"ab_test_winner,omitempty"`

	// Timing
	DelayDays  int `gorm:"default:0" json:"delay_days"`
	DelayHours int `gorm:"default:0" json:"delay_hours"`

	// Conditional Logic
	ConditionType *ConditionType `gorm:"type:text" json:"condition_type,omitempty"`
	ParentStepID  *uuid.UUID     `gorm:"type:uuid" json:"parent_step_id,omitempty"`

	// Performance
	TotalExecuted int `gorm:"default:0" json:"total_executed"`
	TotalSuccess  int `gorm:"default:0" json:"total_success"`
	TotalFailed   int `gorm:"default:0" json:"total_failed"`

	// Timestamps
	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`
}

func (CampaignSequence) TableName() string { return "campaign_sequences" }

func (cs *CampaignSequence) BeforeCreate(tx *gorm.DB) error {
	if cs.ID == uuid.Nil {
		cs.ID = uuid.New()
	}
	return nil
}

// ── CampaignLead ────────────────────────────────────────────────────────────

type CampaignLead struct {
	ID               uuid.UUID          `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	CampaignID       uuid.UUID          `gorm:"type:uuid;not null;index" json:"campaign_id"`
	LeadID           uuid.UUID          `gorm:"type:uuid;not null" json:"lead_id"`
	SenderID         *uuid.UUID         `gorm:"type:uuid" json:"sender_id,omitempty"`
	SenderAccountID  *uuid.UUID         `gorm:"type:uuid" json:"sender_account_id,omitempty"`

	// Lead data (populated via Preload)
	Lead *Lead `gorm:"foreignKey:LeadID" json:"lead,omitempty"`

	// A/B Variant
	Variant *string `gorm:"type:text" json:"variant,omitempty"`

	// Status
	Status            CampaignLeadStatus `gorm:"type:text;default:'pending'" json:"status"`
	CurrentStepID     *uuid.UUID         `gorm:"type:uuid" json:"current_step_id,omitempty"`
	CurrentStepNumber *int               `gorm:"type:int" json:"current_step_number,omitempty"`

	// Engagement
	ConnectionSentAt     *time.Time `gorm:"type:timestamptz" json:"connection_sent_at,omitempty"`
	ConnectionAcceptedAt *time.Time `gorm:"type:timestamptz" json:"connection_accepted_at,omitempty"`
	FirstMessageSentAt   *time.Time `gorm:"type:timestamptz" json:"first_message_sent_at,omitempty"`
	FirstReplyAt         *time.Time `gorm:"type:timestamptz" json:"first_reply_at,omitempty"`
	RepliedAt            *time.Time `gorm:"type:timestamptz" json:"replied_at,omitempty"`
	LastActivityAt       *time.Time `gorm:"type:timestamptz" json:"last_activity_at,omitempty"`

	// Stats
	TotalMessagesSent    int `gorm:"default:0" json:"total_messages_sent"`
	TotalRepliesReceived int `gorm:"default:0" json:"total_replies_received"`

	// Scheduling
	NextActionAt *time.Time `gorm:"type:timestamptz" json:"next_action_at,omitempty"`

	// Error Handling
	ErrorMessage *string `gorm:"type:text" json:"error_message,omitempty"`
	RetryCount   int     `gorm:"default:0" json:"retry_count"`

	// Timestamps
	AddedAt     time.Time  `gorm:"autoCreateTime" json:"added_at"`
	StartedAt   *time.Time `gorm:"type:timestamptz" json:"started_at,omitempty"`
	CompletedAt *time.Time `gorm:"type:timestamptz" json:"completed_at,omitempty"`
	CreatedAt   time.Time  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time  `gorm:"autoUpdateTime" json:"updated_at"`
}

func (CampaignLead) TableName() string { return "campaign_leads" }

func (cl *CampaignLead) BeforeCreate(tx *gorm.DB) error {
	if cl.ID == uuid.Nil {
		cl.ID = uuid.New()
	}
	return nil
}

// ── CampaignActivityLog ─────────────────────────────────────────────────────

type CampaignActivityLog struct {
	ID              uuid.UUID      `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	CampaignID      uuid.UUID      `gorm:"type:uuid;not null;index" json:"campaign_id"`
	CampaignLeadID  *uuid.UUID     `gorm:"type:uuid" json:"campaign_lead_id,omitempty"`
	SequenceStepID  *uuid.UUID     `gorm:"type:uuid" json:"sequence_step_id,omitempty"`
	ActivityType    ActivityType   `gorm:"type:text;not null" json:"activity_type"`
	ActivityStatus  ActivityStatus `gorm:"type:text;not null" json:"activity_status"`
	MessageContent  *string        `gorm:"type:text" json:"message_content,omitempty"`
	ErrorMessage    *string        `gorm:"type:text" json:"error_message,omitempty"`
	Metadata        JSONB          `gorm:"type:jsonb" json:"metadata,omitempty"`
	ExecutedAt      time.Time      `gorm:"autoCreateTime" json:"executed_at"`
	CreatedAt       time.Time      `gorm:"autoCreateTime" json:"created_at"`
}

func (CampaignActivityLog) TableName() string { return "campaign_activity_log" }

func (l *CampaignActivityLog) BeforeCreate(tx *gorm.DB) error {
	if l.ID == uuid.Nil {
		l.ID = uuid.New()
	}
	return nil
}

// ── AccountDailyCounter ─────────────────────────────────────────────────────

type AccountDailyCounter struct {
	ID                uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	LinkedInAccountID uuid.UUID `gorm:"column:linkedin_account_id;type:uuid;not null;index" json:"linkedin_account_id"`
	Date              string    `gorm:"type:date;not null" json:"date"`
	ConnectionsSent   int       `gorm:"default:0" json:"connections_sent"`
	MessagesSent      int       `gorm:"default:0" json:"messages_sent"`
	InMailsSent       int       `gorm:"column:inmails_sent;default:0" json:"inmails_sent"`
	ProfileViews      int       `gorm:"default:0" json:"profile_views"`
	TotalActions      int       `gorm:"default:0" json:"total_actions"`
	CreatedAt         time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt         time.Time `gorm:"autoUpdateTime" json:"updated_at"`
}

func (AccountDailyCounter) TableName() string { return "account_daily_counters" }

func (c *AccountDailyCounter) BeforeCreate(tx *gorm.DB) error {
	if c.ID == uuid.Nil {
		c.ID = uuid.New()
	}
	return nil
}

// ── CampaignWebhook ─────────────────────────────────────────────────────────

type CampaignWebhook struct {
	ID          uuid.UUID   `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	CampaignID  uuid.UUID   `gorm:"type:uuid;not null;index" json:"campaign_id"`
	UserID      uuid.UUID   `gorm:"type:uuid;not null" json:"user_id"`
	URL         string      `gorm:"type:text;not null" json:"url"`
	Secret      *string     `gorm:"type:text" json:"secret,omitempty"`
	Description *string     `gorm:"type:text" json:"description,omitempty"`
	Events      StringArray `gorm:"type:jsonb" json:"events"`
	IsActive    bool        `gorm:"default:true" json:"is_active"`
	CreatedAt   time.Time   `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time   `gorm:"autoUpdateTime" json:"updated_at"`
}

func (CampaignWebhook) TableName() string { return "campaign_webhooks" }

func (w *CampaignWebhook) BeforeCreate(tx *gorm.DB) error {
	if w.ID == uuid.Nil {
		w.ID = uuid.New()
	}
	return nil
}

// ── CampaignWebhookLog ──────────────────────────────────────────────────────

type CampaignWebhookLog struct {
	ID           uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	WebhookID    uuid.UUID `gorm:"type:uuid;not null;index" json:"webhook_id"`
	CampaignID   uuid.UUID `gorm:"type:uuid;not null" json:"campaign_id"`
	Event        string    `gorm:"type:text;not null" json:"event"`
	Success      bool      `gorm:"default:false" json:"success"`
	StatusCode   *int      `gorm:"type:int" json:"status_code,omitempty"`
	ErrorMessage *string   `gorm:"type:text" json:"error_message,omitempty"`
	CreatedAt    time.Time `gorm:"autoCreateTime" json:"created_at"`

	// Relations
	Webhook *CampaignWebhook `gorm:"foreignKey:WebhookID" json:"webhook,omitempty"`
}

func (CampaignWebhookLog) TableName() string { return "campaign_webhook_logs" }

func (l *CampaignWebhookLog) BeforeCreate(tx *gorm.DB) error {
	if l.ID == uuid.Nil {
		l.ID = uuid.New()
	}
	return nil
}

// ── Request / Response DTOs ─────────────────────────────────────────────────

type CreateCampaignInput struct {
	Name                    string               `json:"name" validate:"required"`
	Description             *string              `json:"description,omitempty"`
	LeadListID              *string              `json:"lead_list_id,omitempty"`
	DailyLimit              *int                 `json:"daily_limit,omitempty"`
	Timezone                *string              `json:"timezone,omitempty"`
	SenderIDs               []string             `json:"sender_ids"`
	Sequences               []CreateSequenceInput `json:"sequences"`
	WorkingHoursStart       *string              `json:"working_hours_start,omitempty"`
	WorkingHoursEnd         *string              `json:"working_hours_end,omitempty"`
	WorkingDays             []string             `json:"working_days,omitempty"`
	DelayMinSeconds         *int                 `json:"delay_min_seconds,omitempty"`
	DelayMaxSeconds         *int                 `json:"delay_max_seconds,omitempty"`
	WarmUpEnabled           *bool                `json:"warm_up_enabled,omitempty"`
	WarmUpDays              *int                 `json:"warm_up_days,omitempty"`
	AutoPauseBelowAcceptance *float64            `json:"auto_pause_below_acceptance,omitempty"`
	SkipAlreadyContacted    *bool                `json:"skip_already_contacted,omitempty"`
	StopOnReply             *bool                `json:"stop_on_reply,omitempty"`
}

type CreateSequenceInput struct {
	StepNumber       int     `json:"step_number"`
	StepType         string  `json:"step_type"`
	MessageTemplate  *string `json:"message_template,omitempty"`
	SubjectTemplate  *string `json:"subject_template,omitempty"`
	PostURL          *string `json:"post_url,omitempty"`
	DelayDays        *int    `json:"delay_days,omitempty"`
	DelayHours       *int    `json:"delay_hours,omitempty"`
	ConditionType    *string `json:"condition_type,omitempty"`
	ParentStepID     *string `json:"parent_step_id,omitempty"`
	MessageTemplateB *string `json:"message_template_b,omitempty"`
	ABTestEnabled    *bool   `json:"ab_test_enabled,omitempty"`
	ABTestWinner     *string `json:"ab_test_winner,omitempty"`
}

type UpdateCampaignInput struct {
	Name                    *string  `json:"name,omitempty"`
	Description             *string  `json:"description,omitempty"`
	DailyLimit              *int     `json:"daily_limit,omitempty"`
	Timezone                *string  `json:"timezone,omitempty"`
	Status                  *string  `json:"status,omitempty"`
	WorkingHoursStart       *string  `json:"working_hours_start,omitempty"`
	WorkingHoursEnd         *string  `json:"working_hours_end,omitempty"`
	WorkingDays             []string `json:"working_days,omitempty"`
	DelayMinSeconds         *int     `json:"delay_min_seconds,omitempty"`
	DelayMaxSeconds         *int     `json:"delay_max_seconds,omitempty"`
	WarmUpEnabled           *bool    `json:"warm_up_enabled,omitempty"`
	WarmUpDays              *int     `json:"warm_up_days,omitempty"`
	AutoPauseBelowAcceptance *float64 `json:"auto_pause_below_acceptance,omitempty"`
	SkipAlreadyContacted    *bool    `json:"skip_already_contacted,omitempty"`
	StopOnReply             *bool    `json:"stop_on_reply,omitempty"`
}

// ── Analytics DTOs ──────────────────────────────────────────────────────────

type CampaignAnalyticsFunnel struct {
	TotalLeads  int `json:"total_leads"`
	Sent        int `json:"sent"`
	SentPct     int `json:"sent_pct"`
	Accepted    int `json:"accepted"`
	AcceptedPct int `json:"accepted_pct"`
	Messaged    int `json:"messaged"`
	MessagedPct int `json:"messaged_pct"`
	Replied     int `json:"replied"`
	RepliedPct  int `json:"replied_pct"`
}

type CampaignAnalyticsToday struct {
	ConnectionsSent     int `json:"connections_sent"`
	ConnectionsAccepted int `json:"connections_accepted"`
	MessagesSent        int `json:"messages_sent"`
	RepliesReceived     int `json:"replies_received"`
	DailyLimit          int `json:"daily_limit"`
}

type CampaignAnalyticsTrend struct {
	Date        string `json:"date"`
	Connections int    `json:"connections"`
	Replies     int    `json:"replies"`
}

type CampaignAnalyticsPerStep struct {
	StepNumber int    `json:"step_number"`
	StepType   string `json:"step_type"`
	Sent       int    `json:"sent"`
	Converted  int    `json:"converted"`
	Rate       int    `json:"rate"`
}

type CampaignAnalyticsABResult struct {
	StepID          string  `json:"step_id"`
	VariantASent    int     `json:"variant_a_sent"`
	VariantAReplied int     `json:"variant_a_replied"`
	VariantARate    int     `json:"variant_a_rate"`
	VariantBSent    int     `json:"variant_b_sent"`
	VariantBReplied int     `json:"variant_b_replied"`
	VariantBRate    int     `json:"variant_b_rate"`
	Winner          *string `json:"winner"`
}

type CampaignAnalyticsPerSender struct {
	AccountID       string `json:"account_id"`
	ProfileName     string `json:"profile_name"`
	ConnectionsSent int    `json:"connections_sent"`
	Accepted        int    `json:"accepted"`
	MessagesSent    int    `json:"messages_sent"`
	Replied         int    `json:"replied"`
}

type CampaignAnalytics struct {
	Funnel    CampaignAnalyticsFunnel      `json:"funnel"`
	Today     CampaignAnalyticsToday       `json:"today"`
	Trend     []CampaignAnalyticsTrend     `json:"trend"`
	PerStep   []CampaignAnalyticsPerStep   `json:"per_step"`
	ABResults []CampaignAnalyticsABResult  `json:"ab_results"`
	PerSender []CampaignAnalyticsPerSender `json:"per_sender"`
}

type CampaignStats struct {
	TotalCampaigns       int     `json:"totalCampaigns"`
	ActiveCampaigns      int     `json:"activeCampaigns"`
	TotalLeads           int     `json:"totalLeads"`
	RepliedLeads         int     `json:"repliedLeads"`
	ConnectionAcceptRate float64 `json:"connectionAcceptRate"`
	ReplyRate            float64 `json:"replyRate"`
}

// ── Campaign Template DTOs (static, no DB) ──────────────────────────────────

type CampaignTemplateStep struct {
	StepNumber      int     `json:"step_number"`
	StepType        string  `json:"step_type"`
	MessageTemplate *string `json:"message_template,omitempty"`
	SubjectTemplate *string `json:"subject_template,omitempty"`
	DelayDays       *int    `json:"delay_days,omitempty"`
	DelayHours      *int    `json:"delay_hours,omitempty"`
	ConditionType   *string `json:"condition_type,omitempty"`
}

type CampaignTemplate struct {
	ID                      string                 `json:"id"`
	Name                    string                 `json:"name"`
	Description             string                 `json:"description"`
	StepCount               int                    `json:"step_count"`
	EstimatedAcceptanceRate string                 `json:"estimated_acceptance_rate"`
	Steps                   []CampaignTemplateStep `json:"steps"`
}

// ── Lead Stats DTO ──────────────────────────────────────────────────────────

type CampaignLeadStats struct {
	Total      int `json:"total"`
	Pending    int `json:"pending"`
	InProgress int `json:"in_progress"`
	Paused     int `json:"paused"`
	Completed  int `json:"completed"`
	Failed     int `json:"failed"`
	Removed    int `json:"removed"`
}

File: frontend/src/types/campaign.ts
// ─── Campaign module type definitions ─────────────────────────────────────
// Maps 1:1 with Go backend models in backend/internal/models/campaign.go

// ── Enum types ────────────────────────────────────────────────────────────

export type CampaignStatus = 'draft' | 'active' | 'paused' | 'completed' | 'canceled'

export type CampaignLeadStatus =
  | 'pending'
  | 'in_progress'
  | 'completed'
  | 'replied'
  | 'paused'
  | 'failed'
  | 'removed'
  | 'skipped'

export type SequenceStepType =
  | 'follow'
  | 'like_post'
  | 'connection_request'
  | 'message'
  | 'email'
  | 'inmail'
  | 'view_profile'
  | 'delay'

export type ConditionType =
  | 'accepted'
  | 'not_accepted'
  | 'replied'
  | 'not_replied'
  | 'opened'
  | 'clicked'

export type ActivityType =
  | 'follow'
  | 'unfollow'
  | 'like_post'
  | 'connection_request'
  | 'message_sent'
  | 'reply_received'
  | 'inmail_sent'
  | 'email_sent'
  | 'profile_view'

export type ActivityStatus = 'success' | 'failed' | 'pending' | 'skipped'

// ── Campaign ──────────────────────────────────────────────────────────────

export interface Campaign {
  id: string
  user_id: string
  name: string
  description?: string
  status: CampaignStatus

  // Lead list association
  lead_list_id?: string

  // Configuration
  daily_limit: number
  timezone: string

  // Safety Controls
  working_hours_start?: string
  working_hours_end?: string
  working_days?: string[]
  delay_min_seconds?: number
  delay_max_seconds?: number
  warm_up_enabled?: boolean
  warm_up_days?: number
  warm_up_start_date?: string
  auto_pause_below_acceptance?: number
  skip_already_contacted?: boolean
  stop_on_reply?: boolean

  // Performance Metrics (aggregates)
  total_leads: number
  pending_leads: number
  completed_leads: number
  replied_leads: number

  // Stats
  connection_sent: number
  connection_accepted: number
  messages_sent: number
  replies_received: number
  failed_leads: number

  // Timestamps
  started_at?: string
  paused_at?: string
  completed_at?: string
  created_at: string
  updated_at: string

  // Relations (populated by service layer)
  senders?: CampaignSender[]
  sequences?: CampaignSequence[]
  leads?: CampaignLead[]
}

// ── CampaignSender ────────────────────────────────────────────────────────

export interface CampaignSender {
  id: string
  campaign_id: string
  linkedin_account_id: string
  is_active: boolean
  daily_limit: number

  // Performance Metrics
  leads_assigned: number
  connection_sent: number
  connection_accepted: number
  messages_sent: number
  replies_received: number

  // Timestamps
  created_at: string
  updated_at: string

  // Relations
  linkedin_account?: {
    id: string
    email: string
    profile_name?: string
    profile_picture_url?: string
    status: string
  }
}

// ── CampaignSequence ──────────────────────────────────────────────────────

export interface CampaignSequence {
  id: string
  campaign_id: string
  step_number: number
  step_type: SequenceStepType

  // Action Details
  message_template?: string
  subject_template?: string
  post_url?: string

  // A/B Testing
  message_template_b?: string
  ab_test_enabled?: boolean
  ab_test_winner?: string

  // Timing
  delay_days: number
  delay_hours: number

  // Conditional Logic
  condition_type?: ConditionType
  parent_step_id?: string

  // Performance
  total_executed: number
  total_success: number
  total_failed: number

  // Timestamps
  created_at: string
  updated_at: string
}

// ── CampaignLead ──────────────────────────────────────────────────────────

export interface CampaignLead {
  id: string
  campaign_id: string
  lead_id: string
  sender_id?: string
  sender_account_id?: string

  // Lead data (populated via Preload)
  lead?: {
    id: string
    first_name?: string
    last_name?: string
    full_name?: string
    email?: string
    company?: string
    title?: string
    linkedin_url?: string
  }

  // A/B Variant
  variant?: string

  // Status
  status: CampaignLeadStatus
  current_step_id?: string
  current_step_number?: number

  // Engagement
  connection_sent_at?: string
  connection_accepted_at?: string
  first_message_sent_at?: string
  first_reply_at?: string
  replied_at?: string
  last_activity_at?: string

  // Stats
  total_messages_sent: number
  total_replies_received: number

  // Scheduling
  next_action_at?: string

  // Error Handling
  error_message?: string
  retry_count: number

  // Timestamps
  added_at: string
  started_at?: string
  completed_at?: string
  created_at: string
  updated_at: string
}

// ── CampaignActivityLog ───────────────────────────────────────────────────

export interface CampaignActivityLog {
  id: string
  campaign_id: string
  campaign_lead_id?: string
  sequence_step_id?: string
  activity_type: ActivityType
  activity_status: ActivityStatus
  message_content?: string
  error_message?: string
  metadata?: Record<string, unknown>
  executed_at: string
  created_at: string

  // Populated relation
  campaign_lead?: CampaignLead
}

// ── CampaignWebhook ──────────────────────────────────────────────────────

export interface CampaignWebhook {
  id: string
  campaign_id: string
  user_id: string
  url: string
  secret?: string
  description?: string
  events: string[]
  is_active: boolean
  created_at: string
  updated_at: string
}

// ── CampaignWebhookLog ──────────────────────────────────────────────────

export interface CampaignWebhookLog {
  id: string
  webhook_id: string
  campaign_id: string
  event: string
  success: boolean
  status_code?: number
  error_message?: string
  created_at: string
  webhook?: CampaignWebhook
}

// ── Analytics DTOs ────────────────────────────────────────────────────────

export interface CampaignAnalyticsFunnel {
  total_leads: number
  sent: number
  sent_pct: number
  accepted: number
  accepted_pct: number
  messaged: number
  messaged_pct: number
  replied: number
  replied_pct: number
}

export interface CampaignAnalyticsToday {
  connections_sent: number
  connections_accepted: number
  messages_sent: number
  replies_received: number
  daily_limit: number
}

export interface CampaignAnalyticsTrend {
  date: string
  connections: number
  replies: number
}

export interface CampaignAnalyticsPerStep {
  step_number: number
  step_type: string
  sent: number
  converted: number
  rate: number
}

export interface CampaignAnalyticsABResult {
  step_id: string
  variant_a_sent: number
  variant_a_replied: number
  variant_a_rate: number
  variant_b_sent: number
  variant_b_replied: number
  variant_b_rate: number
  winner?: string
}

export interface CampaignAnalyticsPerSender {
  account_id: string
  profile_name: string
  connections_sent: number
  accepted: number
  messages_sent: number
  replied: number
}

export interface CampaignAnalytics {
  funnel: CampaignAnalyticsFunnel
  today: CampaignAnalyticsToday
  trend: CampaignAnalyticsTrend[]
  per_step: CampaignAnalyticsPerStep[]
  ab_results: CampaignAnalyticsABResult[]
  per_sender: CampaignAnalyticsPerSender[]
}

// ── Campaign Stats ────────────────────────────────────────────────────────

export interface CampaignStats {
  totalCampaigns: number
  activeCampaigns: number
  totalLeads: number
  repliedLeads: number
  connectionAcceptRate: number
  replyRate: number
}

// ── Campaign Template ─────────────────────────────────────────────────────

export interface CampaignTemplateStep {
  step_number: number
  step_type: string
  message_template?: string
  subject_template?: string
  delay_days?: number
  delay_hours?: number
  condition_type?: string
  ab_test_enabled?: boolean
  message_template_b?: string
}

export interface CampaignTemplate {
  id: string
  name: string
  description: string
  icon?: string
  step_count: number
  estimated_acceptance_rate: string
  steps: CampaignTemplateStep[]
}

// ── Campaign Lead Stats ───────────────────────────────────────────────────

export interface CampaignLeadStats {
  total: number
  pending: number
  in_progress: number
  paused: number
  completed: number
  failed: number
  removed: number
}

// ── Request DTOs ──────────────────────────────────────────────────────────

export interface CreateCampaignInput {
  name: string
  description?: string
  lead_list_id?: string
  daily_limit?: number
  timezone?: string
  sender_ids: string[]
  sequences: CreateSequenceInput[]
  working_hours_start?: string
  working_hours_end?: string
  working_days?: string[]
  delay_min_seconds?: number
  delay_max_seconds?: number
  warm_up_enabled?: boolean
  warm_up_days?: number
  auto_pause_below_acceptance?: number
  skip_already_contacted?: boolean
  stop_on_reply?: boolean
}

export interface CreateSequenceInput {
  step_number: number
  step_type: SequenceStepType
  message_template?: string
  subject_template?: string
  post_url?: string
  delay_days?: number
  delay_hours?: number
  condition_type?: ConditionType | string
  parent_step_id?: string
  message_template_b?: string
  ab_test_enabled?: boolean
  ab_test_winner?: string
}

export interface UpdateCampaignInput {
  name?: string
  description?: string
  daily_limit?: number
  timezone?: string
  status?: string
  working_hours_start?: string
  working_hours_end?: string
  working_days?: string[]
  delay_min_seconds?: number
  delay_max_seconds?: number
  warm_up_enabled?: boolean
  warm_up_days?: number
  auto_pause_below_acceptance?: number
  skip_already_contacted?: boolean
  stop_on_reply?: boolean
}

export interface CreateWebhookInput {
  url: string
  secret?: string
  description?: string
  events?: string[]
}
